### ISSUE TITLE
Ensemble modeling problem: missing validation values

### ISSUE THREAD

**Issue by haglad** [biomodhub/biomod2#416]
Hi there biomod team!
I've been working on a SDM for a species of legume, and I ran into an issue with my ensemble results. 
When i run get_evaluations() for my ensemble model output, my 'validation' column appears to have only NA values (as seen below):

thanks for any and all help!
Haley 

this is the code that I am using:
format_bijuga <- 
  BIOMOD_FormatingData(
    resp.var = myResp.bij,
    expl.var = myExpl.bij,
    resp.xy = RespXY.bij,
    resp.name = RespName.bij,
    PA.strategy = 'disk',
    PA.dist.min	= 5000,    
    PA.dist.max	= 100000, #~1 degrees
    PA.nb.absences	= 30, #RF, maxent = PA:P; 
    PA.nb.rep = 10,
    na.rm = T)

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.formated.data -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

dir.name =  .

sp.name =  bijuga

	 29 presences,  0 true absences and  296 undefined points in dataset

	 7 explanatory variables

      bio3            bio11            bio15             bio18       
 Min.   :0.3513   Min.   : 9.898   Min.   :  7.836   Min.   : 312.2  
 1st Qu.:0.6660   1st Qu.:20.649   1st Qu.: 18.869   1st Qu.: 600.1  
 Median :0.8024   Median :24.145   Median : 29.273   Median : 736.6  
 Mean   :0.7563   Mean   :22.675   Mean   : 39.062   Mean   : 800.4  
 3rd Qu.:0.8664   3rd Qu.:25.266   3rd Qu.: 43.243   3rd Qu.: 949.0  
 Max.   :0.9439   Max.   :26.950   Max.   :115.260   Max.   :2110.8  
   phh2o_0_5        variance           elev       
 Min.   :4.608   Min.   :  6906   Min.   :   5.0  
 1st Qu.:5.060   1st Qu.: 26426   1st Qu.:  79.0  
 Median :5.208   Median : 41491   Median : 241.0  
 Mean   :5.292   Mean   : 59301   Mean   : 447.4  
 3rd Qu.:5.444   3rd Qu.: 68210   3rd Qu.: 582.0  
 Max.   :7.164   Max.   :610830   Max.   :2685.0  

 10 Pseudo Absences dataset available ( 
PA1, PA2, PA3, PA4, PA5, PA6, PA7, PA8, PA9, PA10 ) with  
30 (PA1, PA2, PA3, PA4, PA5, PA6, PA7, PA8, PA9, PA10) pseudo absences
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
## Note: i know that i have too many variables for my number of occurrences, I'm hoping to compare my results to an ESM approach ##

cv.b <- bm_CrossValidation(bm.format = format_bijuga,
                           strategy = "block")

base_BiomodOption <- BIOMOD_ModelingOptions(MAXENT = list( path_to_maxent.jar = 'C:/intsia', 
                                                           memory_allocated = 4096))

bmout_bij<- tryCatch(BIOMOD_Modeling(format_bijuga,
                               seed.val = 47,
                                models = c('RF'),    
                                bm.options	= base_BiomodOption,
                                prevalence=0.5,
                                var.import=5, 
                                metric.eval = c('TSS','ROC'),
                                CV.strategy = "user.defined",
                                CV.user.table = cv.b,
                                CV.do.full.models = F,
                                modeling.id = paste('Intsia_bijuga_Jan24',"bmout_bij_12",sep="_"),
                                ))

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.models.out -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Modeling folder : .

Species modeled : bijuga

Modeling id : Intsia_bijuga_Jan24_bmout_bij_12

Considered variables : bio3 bio11 bio15 bio18 phh2o_0_5 variance elev

Computed Models :  bijuga_PA1_RUN1_RF bijuga_PA1_RUN2_RF bijuga_PA1_RUN3_RF 
bijuga_PA1_RUN4_RF bijuga_PA2_RUN1_RF bijuga_PA2_RUN2_RF bijuga_PA2_RUN3_RF 
bijuga_PA2_RUN4_RF bijuga_PA3_RUN1_RF bijuga_PA3_RUN2_RF bijuga_PA3_RUN3_RF 
bijuga_PA3_RUN4_RF bijuga_PA4_RUN1_RF bijuga_PA4_RUN2_RF bijuga_PA4_RUN3_RF 
bijuga_PA4_RUN4_RF bijuga_PA5_RUN1_RF bijuga_PA5_RUN2_RF bijuga_PA5_RUN3_RF 
bijuga_PA5_RUN4_RF bijuga_PA6_RUN1_RF bijuga_PA6_RUN2_RF bijuga_PA6_RUN3_RF 
bijuga_PA6_RUN4_RF bijuga_PA7_RUN1_RF bijuga_PA7_RUN2_RF bijuga_PA7_RUN3_RF 
bijuga_PA7_RUN4_RF bijuga_PA8_RUN1_RF bijuga_PA8_RUN2_RF bijuga_PA8_RUN3_RF 
bijuga_PA8_RUN4_RF bijuga_PA9_RUN1_RF bijuga_PA9_RUN2_RF bijuga_PA9_RUN3_RF 
bijuga_PA9_RUN4_RF bijuga_PA10_RUN1_RF bijuga_PA10_RUN2_RF bijuga_PA10_RUN3_RF 
bijuga_PA10_RUN4_RF

Failed Models :  none

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

modchoseTSS <- eval %>% 
  filter(validation > 0.6) %>% filter(metric.eval == c("TSS"))
modchoseROC <- eval %>% 
  filter(validation > 0.8) %>% filter(metric.eval == c("ROC"))

bmods <- bind_rows(modchoseROC, modchoseTSS)
bmods_v <- bmods$full.name
> bmods_v
[1] "bijuga_PA1_RUN2_RF"  "bijuga_PA4_RUN2_RF"  "bijuga_PA4_RUN4_RF" 
[4] "bijuga_PA7_RUN2_RF"  "bijuga_PA9_RUN4_RF"  "bijuga_PA10_RUN4_RF"
[7] "bijuga_PA8_RUN3_RF"  "bijuga_PA9_RUN4_RF" 

ProjBas <- BIOMOD_Projection(bm.mod = bmout_bij, models.chosen = bmods_v, new.env = myExpl.bij, 
                             new.env.xy = NULL, proj.name = 'baseline_bijuga_SEA_12', metric.binary = NULL,
                             metric.filter = NULL, build.clamping.mask = T, compress = T)

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.projection.out -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Projection directory : ./bijuga/baseline_bijuga_SEA_12

sp.name : bijuga

expl.var.names : bio3 bio11 bio15 bio18 phh2o_0_5 variance elev

modeling.id : Intsia_bijuga_Jan24_bmout_bij_12 ( 
./bijuga/bijuga.Intsia_bijuga_Jan24_bmout_bij_12.models.out )

models.projected : 
bijuga_PA1_RUN2_RF, bijuga_PA4_RUN2_RF, bijuga_PA4_RUN4_RF, bijuga_PA7_RUN2_RF, bijuga_PA9_RUN4_RF, bijuga_PA10_RUN4_RF, bijuga_PA8_RUN3_RF

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

EnsMod <- BIOMOD_EnsembleModeling(bm.mod = bmout_bij, models.chosen = "all" , em.by="all",  
                                  metric.eval = c('TSS', 'ROC'), metric.select.thresh = c(0.7, 0.8), 
                                  metric.select.dataset	= 'validation',
                                  prob.mean = F, prob.median = F, committee.averaging = F, em.algo = c("EMmean"), var.import = 3,
                                  EMwmean.decay = 'proportional', EMci.alpha = 0.05)

-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.ensemble.models.out -=-=-=-=-=-=-=-=-=-=-=-=-=-=

sp.name : bijuga

expl.var.names : bio3 bio11 bio15 bio18 phh2o_0_5 variance elev

models computed: 
bijuga_EMmeanByTSS_mergedData_mergedRun_mergedAlgo, bijuga_EMmeanByROC_mergedData_mergedRun_mergedAlgo

models failed: none

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

get_evaluations(EnsMod)
>get_evaluations(EnsMod)
full.name merged.by.PA merged.by.run merged.by.algo filtered.by   algo
1 bijuga_EMmeanByTSS_mergedData_mergedRun_mergedAlgo   mergedData     mergedRun     mergedAlgo         TSS EMmean
2 bijuga_EMmeanByTSS_mergedData_mergedRun_mergedAlgo   mergedData     mergedRun     mergedAlgo         TSS EMmean
3 bijuga_EMmeanByROC_mergedData_mergedRun_mergedAlgo   mergedData     mergedRun     mergedAlgo         ROC EMmean
4 bijuga_EMmeanByROC_mergedData_mergedRun_mergedAlgo   mergedData     mergedRun     mergedAlgo         ROC EMmean
  metric.eval cutoff sensitivity specificity calibration validation evaluation
1         TSS    529      82.759      79.054       0.618         NA         NA
2         ROC    446      93.103      69.595       0.869         NA         NA
3         TSS    594      96.552      84.122       0.813         NA         NA
4         ROC    597      96.552      85.811       0.941         NA         NA

**Environment Information**
R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 11 x64 (build 22621)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.utf8  LC_CTYPE=English_United Kingdom.utf8   
[3] LC_MONETARY=English_United Kingdom.utf8 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.utf8    

time zone: Europe/London
tzcode source: internal

attached base packages:
 [1] grid      splines   tools     parallel  stats     graphics  grDevices utils    
 [9] datasets  methods   base     

other attached packages:
 [1] sf_1.0-14            latticeExtra_0.6-30  flexsdm_1.3.4       
 [4] automap_1.1-9        shiny_1.8.0          gbm_2.1.8.1         
 [7] usdm_2.1-7           factoextra_1.0.7.999 ade4_1.7-22         
[10] collinear_1.1.1      rstatix_0.7.2        geodata_0.5-9       
[13] viridis_0.6.4        viridisLite_0.4.2    scales_1.2.1        
[16] ENMeval_2.0.4        magrittr_2.0.3       ecospat_4.0.0       
[19] dplyr_1.1.3          tidyr_1.3.0          mecofun_0.5.1       
[22] blockCV_3.1-3        covsel_1.0.0         nsdm_0.0.0.9000     
[25] ggpubr_0.6.0         chron_2.3-61         ROCR_1.0-11         
[28] caret_6.0-94         maxnet_0.1.4         ranger_0.16.0       
[31] lightgbm_4.3.0       RRF_1.9.4            randomForest_4.7-1.1
[34] mgcv_1.9-0           nlme_3.1-163         gam_1.22-3          
[37] foreach_1.5.2        glmnet_4.1-8         Matrix_1.6-1.1      
[40] fst_0.9.8            zoo_1.8-12           rgdal_1.6-7         
[43] writexl_1.4.2        readxl_1.4.3         plyr_1.8.9          
[46] stringr_1.5.1        stringi_1.7.12       data.table_1.14.8   
[49] dismo_1.3-14         rJava_1.0-10         tidyterra_0.5.0     
[52] devtools_2.4.5       usethis_2.2.2        terra_1.7-55        
[55] rasterVis_0.51.6     lattice_0.22-5       gridExtra_2.3       
[58] ggplot2_3.4.4        biomod2_4.2-4        raster_3.6-26       
[61] sp_1.6-0            

loaded via a namespace (and not attached):
  [1] fs_1.6.3               matrixStats_1.1.0      lubridate_1.9.3       
  [4] RColorBrewer_1.1-3     doParallel_1.0.17      profvis_0.3.8         
  [7] backports_1.4.1        utf8_1.2.3             R6_2.5.1              
 [10] urlchecker_1.0.1       withr_2.5.2            prettyunits_1.2.0     
 [13] cli_3.6.1              textshaping_0.3.7      fstcore_0.9.18        
 [16] labeling_0.4.3         sass_0.4.7             proxy_0.4-27          
 [19] systemfonts_1.0.5      parallelly_1.36.0      sessioninfo_1.2.2     
 [22] plotrix_3.8-4          maps_3.4.1.1           rstudioapi_0.15.0     
 [25] FNN_1.1.3.2            generics_0.1.3         shape_1.4.6           
 [28] gtools_3.9.4           car_3.1-2              interp_1.1-4          
 [31] fansi_1.0.5            abind_1.4-5            lifecycle_1.0.4       
 [34] carData_3.0-5          recipes_1.0.8          promises_1.2.1        
 [37] crayon_1.5.2           miniUI_0.1.1.1         cowplot_1.1.1         
 [40] knitr_1.45             pillar_1.9.0           spacetime_1.3-0       
 [43] xgboost_1.7.5.1        future.apply_1.11.0    codetools_0.2-19      
 [46] wk_0.9.0               glue_1.6.2             remotes_2.4.2.1       
 [49] spam_2.10-0            vctrs_0.6.2            png_0.1-8             
 [52] Rdpack_2.5             cellranger_1.1.0       gtable_0.3.4          
 [55] kernlab_0.9-32         TeachingDemos_2.12     cachem_1.0.8          
 [58] xfun_0.41              gower_1.0.1            rbibutils_2.2.16      
 [61] mime_0.12              prodlim_2023.08.28     survival_3.5-7        
 [64] timeDate_4022.108      iterators_1.0.14       hardhat_1.3.0         
 [67] fields_15.2            units_0.8-4            lava_1.7.3            
 [70] ellipsis_0.3.2         Rlof_1.1.3             ipred_0.9-14          
 [73] xts_0.13.1             rprojroot_2.0.4        bslib_0.6.0           
 [76] KernSmooth_2.23-22     rpart_4.1.23           colorspace_2.1-0      
 [79] DBI_1.1.3              nnet_7.3-19            tidyselect_1.2.0      
 [82] processx_3.8.1         compiler_4.3.2         curl_5.1.0            
 [85] mda_0.5-4              gstat_2.1-1            desc_1.4.2            
 [88] classInt_0.4-10        hexbin_1.28.3          corrr_0.4.4           
 [91] callr_3.7.3            PresenceAbsence_1.1.11 plotmo_3.6.2          
 [94] digest_0.6.33          htmltools_0.5.7        pkgconfig_2.0.3       
 [97] jpeg_0.1-10            fastmap_1.1.1          rlang_1.1.1           
[100] htmlwidgets_1.6.3      farver_2.1.1           jquerylib_0.1.4       
[103] jsonlite_1.8.7         ModelMetrics_1.2.2.2   Formula_1.2-5         
[106] dotCall64_1.1-0        s2_1.1.4               patchwork_1.1.3       
[109] munsell_0.5.0          Rcpp_1.0.10            spThin_0.2.0          
[112] pROC_1.18.5            MASS_7.3-60            pkgbuild_1.4.2        
[115] earth_5.3.2            listenv_0.9.0          ggrepel_0.9.4         
[118] deldir_2.0-2           stars_0.6-4            ps_1.7.5              
[121] ggsignif_0.6.4         reshape2_1.4.4         stats4_4.3.2          
[124] pkgload_1.3.3          httpuv_1.6.11          purrr_1.0.1           
[127] future_1.33.0          reshape_0.8.9          broom_1.0.5           
[130] xtable_1.8-4           e1071_1.7-13           later_1.3.1           
[133] class_7.3-22           ragg_1.2.6             intervals_0.15.4      
[136] tibble_3.2.1           memoise_2.0.1          timechange_0.2.0      
[139] globals_0.16.2`

**Comment by HeleneBlt** [comment]
Hello Haley, 

Thanks for posting here 🙏

There is actually no error here 🙂

- At the single models step, you get one value for calibration and one for validation, as you asked to divide each of your PA dataset into 4 blocks (RUN1, RUN2, RUN3 and RUN4).

- But when going to the ensemble model step, you ask to merge all models together, meaning that all repetitions of all PA datasets are combined together. Therefore at this step, calibration and validation lines do not match anymore, and only one evaluation value is calculated over the whole dataset.

- However, you can change this with the parameter `em.by` in `BIOMOD_EnsembleModeling`. For example, you can choose "PA+run" and you will have one model for each PA+run combination. So you can have a validation dataset. But you will not have one unique ensemble model. 

Please, do not hesitate if this is not clear and you need more precisions.

Hélène

**Comment by haglad** [comment]
That makes a lot of sense, thank you very much!
